service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -1085,6 +1087,26 @@ protected void onCommandConnectionEstablished(final HonoConnection commandConnec
                 .map(registrationAssertion);
     }
 
+    /**
+     * Checks whether this adapter supports mapping of incoming command messages to the gateway that the target device
+     * is connected to. If not, the {@link #updateLastGateway(JsonObject, String, String, Device, SpanContext)} method
+     * will associate the device id with itself instead of with the gateway, causing the mapping to be disabled.
+     * <p>
+     * This default implementation returns a succeeded future with {@code true} as value here.
+     * <p>
+     * May be overridden by sub-classes to disable the gateway mapping for this adapter.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device to update the last known gateway for.
+     * @param authenticatedDevice The device that has authenticated to this protocol adapter.
+     * @return A succeeded future indicating the outcome of the operation. Its value will be {@code true} if gateway
+     *         mapping is enabled for this adapter.
+     */
+    protected Future<Boolean> isGatewayMappingEnabled(final String tenantId, final String deviceId,
```

Comment:
how about declaring this *abtract* and adding final implementations to the standard adapters in order to prevent users from overriding it, which would make it harder to remove the method again once we have #1272 in place?

Reply:
Making it _abstract_ would impose the requirement to override the method for non-standard adapters. I would rather not have that (makes it also harder removing the method again).

As an alternative, we could therefore keep the method but still override it with final methods in the standard adapters (in accordance to the protocol adapter methods where the command consumer gets created usually being final as well).

Reply:
I've amended the commit accordingly (omitting `VertxBasedAmqpProtocolAdapter` since that class is final anyway).